DNF5 prototype

pkratoch opened the pull request.

pep8speaks commented:
Hello @pkratoch! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`pyanaconda/modules/payloads/payload/dnf/dnf_manager.py`](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/pyanaconda/modules/payloads/payload/dnf/dnf_manager.py):

> [Line 572:9](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/pyanaconda/modules/payloads/payload/dnf/dnf_manager.py#L572): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 588:9](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/pyanaconda/modules/payloads/payload/dnf/dnf_manager.py#L588): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 684:13](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/pyanaconda/modules/payloads/payload/dnf/dnf_manager.py#L684): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 847:13](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/pyanaconda/modules/payloads/payload/dnf/dnf_manager.py#L847): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '







* In the file [`pyanaconda/modules/payloads/payload/dnf/transaction_progress.py`](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/pyanaconda/modules/payloads/payload/dnf/transaction_progress.py):

> [Line 90:13](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/pyanaconda/modules/payloads/payload/dnf/transaction_progress.py#L90): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 95:9](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/pyanaconda/modules/payloads/payload/dnf/transaction_progress.py#L95): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 115:13](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/pyanaconda/modules/payloads/payload/dnf/transaction_progress.py#L115): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 167:100](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/pyanaconda/modules/payloads/payload/dnf/transaction_progress.py#L167): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 179:100](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/pyanaconda/modules/payloads/payload/dnf/transaction_progress.py#L179): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)
> [Line 194:100](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/pyanaconda/modules/payloads/payload/dnf/transaction_progress.py#L194): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (100 > 99 characters)















* In the file [`tests/unit_tests/pyanaconda_tests/modules/payloads/payload/test_module_payload_dnf_manager.py`](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/tests/unit_tests/pyanaconda_tests/modules/payloads/payload/test_module_payload_dnf_manager.py):

> [Line 31:1](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/tests/unit_tests/pyanaconda_tests/modules/payloads/payload/test_module_payload_dnf_manager.py#L31): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 32:1](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/tests/unit_tests/pyanaconda_tests/modules/payloads/payload/test_module_payload_dnf_manager.py#L32): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 33:1](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/tests/unit_tests/pyanaconda_tests/modules/payloads/payload/test_module_payload_dnf_manager.py#L33): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 34:1](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/tests/unit_tests/pyanaconda_tests/modules/payloads/payload/test_module_payload_dnf_manager.py#L34): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 35:1](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/tests/unit_tests/pyanaconda_tests/modules/payloads/payload/test_module_payload_dnf_manager.py#L35): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 36:1](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/tests/unit_tests/pyanaconda_tests/modules/payloads/payload/test_module_payload_dnf_manager.py#L36): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 37:1](https://github.com/rhinstaller/anaconda/blob/c6a7c0de426a56bb72c1ae5624faa8ad9a7ce569/tests/unit_tests/pyanaconda_tests/modules/payloads/payload/test_module_payload_dnf_manager.py#L37): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

pkratoch commented:
This dnf5 PR is needed for the prototype to finish the installation: https://github.com/rpm-software-management/dnf5/pull/1697

pkratoch commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -88,21 +85,60 @@ class InvalidSelectionError(DNFManagerError):
     """The software selection couldn't be resolved."""
 
 
+class DNFConfigWrapper(object):
+    """This is a temporary wrapper of a DNF config object."""
+
+    def __init__(self, config):
+        """Wrap the DNF config object."""
+        self._config = config
+
+    def __getattr__(self, name):
+        """Get the attribute.
+
+        Called when an attribute lookup has not found
+        the attribute in the usual places.
+        """
+        option = getattr(self._config, name)()
+        return option.get_value()
+
+    def __setattr__(self, name, value):
+        """Set the attribute.
+
+        Called when an attribute assignment is attempted.
+        """
+        if name in ["_config"]:
+            return super().__setattr__(name, value)
+
+        option = getattr(self._config, name)()
+        option.set(value)
```
Regarding the `DNFConfigWrapper`, right now, it is quite ugly, since the option names are in the form `get_<name>_option`. I have two possible solutions:

1. Remove the wrapper and access the options directly. This would look like this:
    ```
    config.get_cachedir_option().get_value()
    config.get_cachedir_option().set(DNF_CACHE_DIR)
    ```

2. Keep the wrapper, but use simple `<name>` form in anaconda and wrap it with `"get_" + name + "_option"` in the `DNFConfigWrapper`.

The second solution would mean more readable access to the options, but the disadvantage is that it relies on a specific naming scheme in libdnf5. This could become a problem if we needed something else from the config e.g. to load config from parser (`ConfigMain::load_from_parser`), but it's not needed now. I prefer the second solution, but I don't know why the `DNFConfigWrapper` is described as temporary and what was the plan for future.

pkratoch replied:
I changed the wrapper to contain all the necessary options as properties. It got quite long, so let me know if it's acceptable. Maybe it could be moved to a separate file? Alternatively, I will change it to remove the wrapper entirely.

pkratoch commented:
I also pushed 2 commits changing the mocking in unit tests. Sorry I didn't notice it sooner that the tests do not pass. I am not totally sure about the second one that adds the try block, though. It's ok to do it like that in the test, but I am not sure if we can rely on dnf always sending the quit or error callback. If dnf crashed and didn't send anything, anaconda would wait forever. But I don't think there is anything that can be done on anaconda side.

Also, I put the commits on top of the branch, but I would like to eventually amend the relevant commits in history, so that the history is not so chaotic. Let me know if it's ok to do that.

KKoukiou commented:
> I also pushed 2 commits changing the mocking in unit tests. Sorry I didn't notice it sooner that the tests do not pass. I am not totally sure about the second one that adds the try block, though. It's ok to do it like that in the test, but I am not sure if we can rely on dnf always sending the quit or error callback. If dnf crashed and didn't send anything, anaconda would wait forever. But I don't think there is anything that can be done on anaconda side.
> 
> Also, I put the commits on top of the branch, but I would like to eventually amend the relevant commits in history, so that the history is not so chaotic. Let me know if it's ok to do that.

Sure feel free to amend.

jan-kolarik commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -88,21 +85,265 @@ class InvalidSelectionError(DNFManagerError):
+
+    @property
+    def baseurl(self):
+        return self._config.get_baseurl_option().get_value()
```
This seems not necessary as dnf5 is already providing these attributes in Python bindings. Check [the implementation](https://github.com/rpm-software-management/dnf5/blob/5.2.6.2/bindings/libdnf5/conf.i#L118) in our SWIG interface file and e.g. [this test](https://github.com/rpm-software-management/dnf5/blob/5.2.6.2/test/python3/libdnf5/conf/test_option.py#L59).

pkratoch replied:
Oh, wow, how did I miss that? It was even done in the same PR as the renaming, but I guess I focused too much on how the wrapped had been done before that I didn't notice. This is really neat, I will change the PR to drop the wrapper and use the attributes directly.

jan-kolarik replied:
I think this is the case, when presenting a simple demo of new features added to our projects during the sprint review or retrospective would be useful. :slightly_smiling_face:

M4rtinK commented:
@pkratoch Looks like it would be good to rebase the branch - there are quite a few conflicts by this point + we also switched to using `main` as primary upstream branch just before the holidays. Thanks in advance! :)

pkratoch commented:
> @pkratoch Looks like it would be good to rebase the branch - there are quite a few conflicts by this point + we also switched to using `main` as primary upstream branch just before the holidays. Thanks in advance! :)

Rebased.

I also finished the fix of download callbacs and the dnf5 version that is necessary for it to work is now in rawhide already. :)

M4rtinK commented:
Seeing a really good progress with the latest code - the issue with environment & group listing seems to be fixed now. :)

I've built a comparison image (eq. `main` branch & similar Rawhide packages) & noted a few smaller differences in behavior:

![software_selection_dnf4](https://github.com/user-attachments/assets/eb58deab-e4c7-4b09-a955-f000acd65f0d)
`main` branch/DNF 4

![software_selection_dnf5](https://github.com/user-attachments/assets/63fcd5e5-f179-4df5-8436-f249288a38f0)
DNF 5 PR

Looks like the DNF 5 results are sorted differently ? This is actually quite important, as IIRC the default environment is the one on top & the comps are set up that way on purpose. So I think this should be fixed to get the same ordering. :)

Then there is a small difference in progress reporting during package installation:

![installing_dnf4](https://github.com/user-attachments/assets/cc196928-307a-4666-b939-825afea2cf92)
`main` branch/DNF 4

![installing_dnf5](https://github.com/user-attachments/assets/a6a07708-20b5-4b74-b0f0-7e308f12c162)
DNF 5 PR

Note the (package number/number of packages) that is missing from the DNF 5 version. While definitely not fatal, I think its an important indicator of progress or being made or even the process getting stuck.

Interestingly, there is no numeric progress indicator for the configuration phase on both DNF 4 & DNF 5:

![configuring_dnf4](https://github.com/user-attachments/assets/2399369b-559d-4f4e-9ce6-36b0ba98eb67)
`main` branch/DNF 4

![configuring_dnf5](https://github.com/user-attachments/assets/84131fa3-ff6a-491f-a71a-710f5d46da3b)
DNF 5 PR

Maybe we could add it when we are at it (if technically possible) to be more consistent ?

I've also tried booting the installed system & all seemed fine. :)

M4rtinK commented:
(You can either check the test definitions in the kickstart test repo or go the the test run here on github, download the `logs.zip` file & check the `logs/kickstart-tests/data/logs/kstest-list-substituted` folder inside, as that should have all the *compiled* kickstarts as they have been run.)

First failed kickstart test is `packages-ignoremissing` - I think the relevant part of the test is this:

```
%packages --ignoremissing
fake-package-name
%end
```

The test is checking if the `--ignoremissing` flag for the `%packages` section works. If this flag is passed, unknown/missing packages mentioned in the `%packages` section will be ignored.

This does not seem to work at the moment with the DNF 5 payload and it fails due to `fake-package-name` not being available.


```
12:57:23,319 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.requirements:Applying requirements: [Requirement(name='btrfs-progs', reason='Required to manage storage devices.', type='package'), Requirement(name='e2fsprogs', reason='Required to manage storage devices.', type='package'), Requirement(name='grub2', reason='Necessary for the bootloader configuration.', type='package'), Requirement(name='grub2-tools', reason='Necessary for the bootloader configuration.', type='package'), Requirement(name='chrony', reason='Needed to run NTP service.', type='package'), Requirement(name='langpacks-en', reason="Required to support the locale 'en_US.UTF-8'.", type='package')]
12:57:23,319 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.requirements:Requirement 'btrfs-progs' is applied. Reason: Required to manage storage devices.
12:57:23,319 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.requirements:Requirement 'e2fsprogs' is applied. Reason: Required to manage storage devices.
12:57:23,319 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.requirements:Requirement 'grub2' is applied. Reason: Necessary for the bootloader configuration.
12:57:23,319 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.requirements:Requirement 'grub2-tools' is applied. Reason: Necessary for the bootloader configuration.
12:57:23,320 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.requirements:Requirement 'chrony' is applied. Reason: Needed to run NTP service.
12:57:23,320 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.requirements:Requirement 'langpacks-en' is applied. Reason: Required to support the locale 'en_US.UTF-8'.
12:57:23,320 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.validation:Resolving the software selection.
12:57:23,321 WARNING org.fedoraproject.Anaconda.Modules.Payloads:INFO:anaconda.modules.payloads.payload.dnf.dnf_manager:Including specs: ['@core', 'fake-package-name', 'kernel', 'btrfs-progs', 'e2fsprogs', 'grub2', 'grub2-tools', 'chrony', 'langpacks-en']
12:57:23,321 WARNING org.fedoraproject.Anaconda.Modules.Payloads:INFO:anaconda.modules.payloads.payload.dnf.dnf_manager:Excluding specs: []
12:57:23,321 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.dnf_manager:Resolving the software selection.
12:57:23,491 WARNING org.fedoraproject.Anaconda.Modules.Payloads:INFO:anaconda.modules.payloads.payload.dnf.dnf_manager:The software selection has been resolved (409 packages selected).
12:57:23,492 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.dnf_manager:Resolving has been completed: ValidationReport(error_messages=[], warning_messages=['No match for argument: fake-package-name'])
12:57:23,493 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.validation:Resolving has been completed: ValidationReport(error_messages=[], warning_messages=['No match for argument: fake-package-name'])
12:57:23,493 WARNING org.fedoraproject.Anaconda.Modules.Payloads:WARNING:anaconda.modules.payloads.payload.dnf.installation:The packages were resolved with warnings:
12:57:23,493 WARNING org.fedoraproject.Anaconda.Modules.Payloads:No match for argument: fake-package-name
12:57:23,493 WARNING org.fedoraproject.Anaconda.Modules.Payloads:INFO:anaconda.core.threads:Thread Failed: AnaTaskThread-ResolvePackagesTask-2 (140107733325504)
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:ERROR:anaconda.modules.common.task.task:Thread AnaTaskThread-ResolvePackagesTask-2 has failed: Traceback (most recent call last):
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/core/threads.py", line 281, in run
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    threading.Thread.run(self)
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~~~~~~~~~^^^^^^
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/threading.py", line 992, in run
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self._target(*self._args, **self._kwargs)
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 97, in _thread_run_callback
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self._task_run_callback()
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~~~~~~~~~~~~^^
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 110, in _task_run_callback
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self._set_result(self.run())
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:                     ~~~~~~~~^^
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/payloads/payload/dnf/installation.py", line 152, in run
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    raise NonCriticalInstallationError(message)
12:57:23,507 WARNING org.fedoraproject.Anaconda.Modules.Payloads:pyanaconda.modules.common.errors.installation.NonCriticalInstallationError: No match for argument: fake-package-name
12:57:23,508 WARNING org.fedoraproject.Anaconda.Modules.Payloads:INFO:anaconda.core.threads:Thread Done: AnaTaskThread-ResolvePackagesTask-2 (140107733325504)
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:WARNING:dasbus.server.handler:The call org.fedoraproject.Anaconda.Task.Finish has failed with an exception:
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:Traceback (most recent call last):
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib/python3.13/site-packages/dasbus/server/handler.py", line 455, in _method_callback
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    result = self._handle_call(
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:        interface_name,
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ...<2 lines>...
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:        **additional_args
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    )
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib/python3.13/site-packages/dasbus/server/handler.py", line 265, in _handle_call
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    return handler(*parameters, **additional_args)
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task_interface.py", line 115, in Finish
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self.implementation.finish()
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~~~~~~~~~~~~~~~^^
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 176, in finish
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    thread_manager.raise_if_error(self._thread_name)
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/core/threads.py", line 172, in raise_if_error
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    raise exc_info[1]
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/core/threads.py", line 281, in run
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    threading.Thread.run(self)
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~~~~~~~~~^^^^^^
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/threading.py", line 992, in run
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self._target(*self._args, **self._kwargs)
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 97, in _thread_run_callback
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self._task_run_callback()
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~~~~~~~~~~~~^^
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 110, in _task_run_callback
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self._set_result(self.run())
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:                     ~~~~~~~~^^
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/payloads/payload/dnf/installation.py", line 152, in run
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    raise NonCriticalInstallationError(message)
12:57:23,598 WARNING org.fedoraproject.Anaconda.Modules.Payloads:pyanaconda.modules.common.errors.installation.NonCriticalInstallationError: No match for argument: fake-package-name
```

The `packages-and-groups-ignoremissing` test is failing for the same reason, it just tests also for missing groups.

M4rtinK commented:
Next one is `fedora-live-image-build` - this is basically a snapshot of a kickstart used to build a Fedora live image. The idea is to roughly check if Live image builds based on Anaconda still work.

These kickstarts are manually updated once in a while. Also it looks like Fedora might be switching to some other tool (Kiwi ?) that does not use Anaconda for Live image tools in the future, so maybe we could drop this test in the future ? In any case, it does not work right now.

```
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:ERROR:anaconda.modules.common.task.task:Thread AnaTaskThread-ResolvePackagesTask-2 has failed: Traceback (most recent call last):
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/core/threads.py", line 281, in run
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    threading.Thread.run(self)
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~~~~~~~~~^^^^^^
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/threading.py", line 992, in run
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self._target(*self._args, **self._kwargs)
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 97, in _thread_run_callback
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self._task_run_callback()
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~~~~~~~~~~~~^^
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 110, in _task_run_callback
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self._set_result(self.run())
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:                     ~~~~~~~~^^
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/payloads/payload/dnf/installation.py", line 152, in run
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    raise NonCriticalInstallationError(message)
12:52:45,190 WARNING org.fedoraproject.Anaconda.Modules.Payloads:pyanaconda.modules.common.errors.installation.NonCriticalInstallationError: Argument 'gfs2-utils' matches only excluded packages.
12:52:45,193 WARNING org.fedoraproject.Anaconda.Modules.Payloads:INFO:anaconda.core.threads:Thread Done: AnaTaskThread-ResolvePackagesTask-2 (140081558005440)
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:WARNING:dasbus.server.handler:The call org.fedoraproject.Anaconda.Task.Finish has failed with an exception:
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:Traceback (most recent call last):
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib/python3.13/site-packages/dasbus/server/handler.py", line 455, in _method_callback
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    result = self._handle_call(
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:        interface_name,
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ...<2 lines>...
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:        **additional_args
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    )
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib/python3.13/site-packages/dasbus/server/handler.py", line 265, in _handle_call
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    return handler(*parameters, **additional_args)
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task_interface.py", line 115, in Finish
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self.implementation.finish()
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~~~~~~~~~~~~~~~^^
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 176, in finish
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    thread_manager.raise_if_error(self._thread_name)
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/core/threads.py", line 172, in raise_if_error
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    raise exc_info[1]
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/core/threads.py", line 281, in run
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    threading.Thread.run(self)
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~~~~~~~~~^^^^^^
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/threading.py", line 992, in run
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self._target(*self._args, **self._kwargs)
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 97, in _thread_run_callback
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self._task_run_callback()
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    ~~~~~~~~~~~~~~~~~~~~~~~^^
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/common/task/task.py", line 110, in _task_run_callback
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    self._set_result(self.run())
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:                     ~~~~~~~~^^
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:  File "/usr/lib64/python3.13/site-packages/pyanaconda/modules/payloads/payload/dnf/installation.py", line 152, in run
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:    raise NonCriticalInstallationError(message)
12:52:45,275 WARNING org.fedoraproject.Anaconda.Modules.Payloads:pyanaconda.modules.common.errors.installation.NonCriticalInstallationError: Argument 'gfs2-utils' matches only excluded packages.
```

The (rather long) kickstart used can be seen here: https://github.com/rhinstaller/kickstart-tests/blob/master/fedora-live-image-build.ks.in

The error complains about `gfs2-utils` being excluded, which it indeed is:

```
%packages
@^workstation-product-environment
@anaconda-tools
aajohan-comfortaa-fonts
anaconda
anaconda-install-env-deps
anaconda-live
chkconfig
dracut-live
glibc-all-langpacks
initscripts
kernel
kernel-modules
kernel-modules-extra
-@dial-up
-@input-methods
-@standard
-device-mapper-multipath
-fcoe-utils
-gfs2-utils
-reiserfs-utils

%end
```

That is correct & expected. The final [spin kickstarts](https://pagure.io/fedora-kickstarts) still used to build Live images are compiled from individual kickstart snippets, where individual spins build on a common base & often cut id down via excludes like this. IIRC @pkratoch mentioned there are currently issues with package and group level excludes in DNF 5 & I believe that is what we are seeing here.

M4rtinK commented:
Looking at `repo-enable` - this enables RPM repositories using the `repo` kickstart command with the `--name` option:

https://github.com/rhinstaller/kickstart-tests/blob/master/repo-enable.ks.in

```
# Enable the repository in the installation environment.
repo --name=addon-a
```

There are two repo files deployed (a & b) and the test enables just repo a, then check just packages from the a repo are installed.

According to log, it looks like no packages were installed:

`*** mandatory-package-from-addon-a was not installed`

This I guess implies `repo --name` does not do anything at the moment ?

M4rtinK commented:
For `repo-exclude` we are trying to enable 2 repos by name (a & b) & then exclude a package from from the a repo via the repo command:

https://github.com/rhinstaller/kickstart-tests/blob/master/repo-exclude.ks.in

Important bit being:

```
repo --name=addon-a --baseurl=REPO_A_URL --cost=25 --excludepkgs=package-1
repo --name=addon-b --baseurl=REPO_B_URL --cost=50

%packages
package-1
package-2
optional-package-from-addon-a
optional-package-from-addon-b
%end
```

As exepcted (eq. `repo --name` not working) the repos are not enabled and test fails due to no packages being installed from the repositories:

```
*** package mandatory-package-from-addon-a was not installed
*** package mandatory-package-from-addon-b was not installed
```

M4rtinK commented:
The `repo-include` test tests the `--cost` and `--includepkgs` options of the repo command - it also is broken due to `repo --name` not working.

M4rtinK commented:
Unlike the other tests so far, the `groups-and-envs-2` test timed out. 

As for what this test does - its a catch-all related groups, environments and related (more) advanced features:

https://github.com/rhinstaller/kickstart-tests/blob/master/groups-and-envs-2.ks.in

The relevant part:

```
# what we are testing there:
# - that we can exclude groups which are a part of an environment
# - that we can exclude groups we have specified in ourselves
#   (this could be important for multiple ksincluded %packages sections)
# - that the --optional flag for package groups is working
# - that the --nodefaults flag for package groups is working

%ksappend repos/default.ks
%ksappend common/common_no_payload.ks

%packages
# (1) Test that you can remove a group that's part of an environment.
@^xfce-desktop-environment
-@dial-up

# (2) Test that you can add and then remove a group.
@3d-printing
-@3d-printing

# (3) Test that --optional works.
@container-management --optional

# (4) Test that --nodefaults works.
@rpm-development-tools --nodefaults
%end
```

Checking the `virt-install.log` file from the test, it seems to get stuck on:

```
13:02:11,236 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.dnf_manager:Resolving has been completed: ValidationReport(error_messages=['No match for argument: container-management/mandatory,default,conditional,optional', 'No match for argument: rpm-development-tools/mandatory,conditional'], warning_messages=[])
13:02:11,237 WARNING org.fedoraproject.Anaconda.Modules.Payloads:DEBUG:anaconda.modules.payloads.payload.dnf.validation:Resolving has been completed: ValidationReport(error_messages=['No match for argument: container-management/mandatory,default,conditional,optional', 'No match for argument: rpm-development-tools/mandatory,conditional'], warning_messages=[])
13:02:11,239 WARNING org.fedoraproject.Anaconda.Modules.Payloads:INFO:anaconda.core.threads:Thread Done: AnaTaskThread-CheckPackagesSelectionTask-1 (139704025712320)
13:02:11,287 INFO anaconda:anaconda: lifecycle: Module initialized: PasswordSpoke
13:02:11,370 DEBUG anaconda:anaconda: payload.dnf.payload: The selection has been checked: ValidationReport(error_messages=['No match for argument: container-management/mandatory,default,conditional,optional', 'No match for argument: rpm-development-tools/mandatory,conditional'], warning_messages=[])
```

So it looks like the `@rpm-development-tools --nodefaults` line from the `%packages` section somehow got transformed to `rpm-development-tools/mandatory,conditional` and `@container-management --optional` got turned to `container-management/mandatory,default,conditional,optional`. 

Apparently those are not valid package/group/env specs DNF 5 can understand, so software configuration validation fails, automated installation is interrupted & waits for user input, until it times out.

To summarize, I think this test managed to uncover that handling of the `--nodefaults` and `--optional` flags that can be passed to specs in the `%packages` section is currently broken.

More about the section and spec level options can be found here: 

https://pykickstart.readthedocs.io/en/latest/kickstart-docs.html#chapter-9-package-selection

M4rtinK commented:
/kickstart-test --skip-testtypes whatever

pkratoch commented:
Hi, @M4rtinK, I pushed a commit which adds the progress during package installation (the `(installed/total)` numbers). I am not adding it for the configuration, because that's just callbacks for scriptlets running and it's not actually a phase, since some scriptlets (pre-install) are running in between the package installs. Also, there is no progress to report coming from dnf.

I have yet to look into the comps order and kickstart tests results.

M4rtinK commented:
> Hi, @M4rtinK, I pushed a commit which adds the progress during package installation (the `(installed/total)` numbers). I am not adding it for the configuration, because that's just callbacks for scriptlets running and it's not actually a phase, since some scriptlets (pre-install) are running in between the package installs. Also, there is no progress to report coming from dnf.

That's totally fine - I was wondering why it was not implemented before for the configuration phase & I guess it was likely similar architectural limitations. :)

> 
> I have yet to look into the comps order and kickstart tests results.

ACK :)

pkratoch commented:
Regarding the `repo-*` tests, I think the root cause of their failures is actually, surprisingly, that dnf5 cannot yet merge group definitions from multiple repositories: https://github.com/rpm-software-management/dnf5/issues/183

The `optional-package-from-addon-a` is installed, because it's explicitely mentioned in the `%packages` section, but the `mandatory-package-from-addon-a` comes from group `@core` defined in the additional repo, but the same group is also in the Fedora repository and only one definition (from the Fedora repo) is taken.

pkratoch commented:
Added a commit with adding group and environment excludes, but it requires https://github.com/rpm-software-management/dnf5/pull/2107 .

M4rtinK commented:
/kickstart-test --skip-testtypes whatever

M4rtinK commented:
/build-image --boot.iso

M4rtinK commented:
/kickstart-test --skip-testtypes whatever

rvykydal commented on tests/unit_tests/pyanaconda_tests/modules/payloads/payload/test_dnf_initialization.py:
```diff
@@ -262,16 +262,17 @@ def test_treeinfo_repositories(self):
             # The treeinfo repositories are configured.
             assert dnf_manager.enabled_repositories == [
-                "anaconda", "AppStream"
+                "AppStream", "anaconda"
```
I see, the list is sorted.

KKoukiou commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -661,27 +783,33 @@ def install_packages(self, callback, timeout=20):
 
     @staticmethod
-    def _run_transaction(base, display):
+    def _run_transaction(base, transaction, progress):
```
Do you need to pass transaction?  Can't you just access the class _transaction variable from within the method?

pkratoch replied:
Probably not. I will take a look at it, but this will likely affect tests, so it's not as easy as to just remove it.

pkratoch replied:
So, I just added the parameter description to the docstring.

KKoukiou commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -616,17 +709,46 @@ def download_packages(self, callback):
+
+        return [
+            tspkg.get_package() for tspkg in self._transaction.get_transaction_packages()
+            if libdnf5.base.transaction.transaction_item_action_is_inbound(tspkg.get_action())
```
What is 'if libdnf5.base.transaction.transaction_item_action_is_inbound(tspkg.get_action())' doing?

pkratoch replied:
I added a comment that should explain that. I added it in an extra commit, because this function was completely written by Vendy and I don't want to alter her initial commit.

KKoukiou commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -91,33 +91,29 @@ class MetadataError(DNFManagerError):
         self._download_location = None
         self._md_hashes = {}
         self._enabled_system_repositories = []
+        self._repositories_loaded = False
```
Why do we need this safeguard now, whereas before it was not needed? Is load_repository aka configure_repository called more than once now?

pkratoch replied:
Because previously we checked other things, such as if `_base.sack` exists, but it's not possible in dnf5. There is a `repos_updated_and_loaded` variable, but it's private.

pkratoch commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -227,24 +253,42 @@ def default_environment(self):
 
         return None
 
+    @property
```
The way that groups and environments are stored is one of the two temporary things (because of https://github.com/rpm-software-management/dnf5/issues/1530 ). I agree this must be changed but I will probably not fix this right away, but think first if we can fix the dnf issue.

rvykydal commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -734,26 +880,30 @@ def add_repository(self, data: RepoConfigurationData):
+                repo = self._create_repository(data)
+            else:
+                # Replace the existing repository with a new one.
+                repo = self._configure_repository(repositories.get(), data)
```
Replace may be a bit misleading here and in the doc string given what is actually done internally. Maybe 'update configuration' / 'reconfigure'?

pkratoch replied:
Thanks, I changed the comment and the method docstring.

rvykydal reviewed (approved):
Looks good to me. Thank you!

KKoukiou commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -127,77 +123,111 @@ def _base(self):
 
         return self.__base
 
+    @property
```
I believe in anaconda codebase there is unwritten rule that @properties are readonly. Here you are doing a lazy initialization of this property, therefore it's not readonly. 

I suggest, to either create a setter function or use https://www.geeksforgeeks.org/python-functools-cached_property/ in order for the None checks to be skipped.

@M4rtinK do you know if we have a convention for this?

KKoukiou commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -494,105 +567,131 @@ def is_package_available(self, package_spec):
+                libdnf5.base.GoalProblem_MODULE_SOLVER_ERROR | \
+                libdnf5.base.GoalProblem_MODULE_CANNOT_SWITH_STREAMS
+
+            # If we didn't already run with skip_unavailable=True and the problems are not from the
```
Does this override the user or system configuration, even though the config explicitly has skip_unavailable = False?

M4rtinK replied:
This behavior is actually quite necessary on RHEL for example, where for whatever reason the insights-client package is usually not available until later in thy cycle & we enable Insights by default as was requested.

So with this feature you can skip the missing Insights package and finish the installation. If we aborted installation due to the missing package, it could break some RHEL testing. If we silently ignored it, we could miss Insights missing at GA or similar.

KKoukiou commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -127,77 +123,111 @@ def _base(self):
+
+        return self.__goal
+
+    @property
```
Can we just have a single goal, aka the original, and setter for that? 

in this way, we can try once, then set again the goal, and try again?

pkratoch replied:
If you mean to replace the `_goal` with `_goal_skip_unavailable` only when the first `_goal` doesn't succeed, then it isn't possible. We need both at the beginning, because both must be populated in the `apply_specs` method. We do it the same way in dnf5 e.g. when resolving modules (see https://github.com/rpm-software-management/dnf5/blob/main/libdnf5/module/module_sack.cpp#L477-L484 where four goals are used).

pkratoch commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -280,41 +328,60 @@ def _get_environment_data(self, env) -> CompsEnvironmentData:
+            log.warning("There is no metadata about environments and groups!")
+            return []
+        if self._query_groups is None:
+            self._query_groups = libdnf5.comps.GroupQuery(self._base)
```
`Environment.get_groups` is to get only groups that are part of a given environment. `GroupQuery` is used here, because the method gets all the available groups. Environments have grouplist and optionlist (list of groups that are optional for that environment), but groups are not optional/non-optional on their own.

pkratoch commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -494,105 +567,131 @@ def is_package_available(self, package_spec):
+
+        log.info("Excluding package specs: %s", package_excludes)
+        excludes = libdnf5.rpm.PackageQuery(self._base)
+        excludes.filter_name(package_excludes, libdnf5.common.QueryCmp_GLOB)
```
Really? I thought that globs are allowed in the kickstart... I'm not sure what the official documentation is, but e.g. here: https://docs.fedoraproject.org/en-US/fedora/f36/install-guide/appendixes/Kickstart_Syntax_Reference/
"You can use the asterisk character (*) as a wildcard in package names." The same is in the example with groups.

M4rtinK replied:
Yeah, I think globs are expected to work.

pkratoch commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -280,41 +328,60 @@ def _get_environment_data(self, env) -> CompsEnvironmentData:
+        :return libdnf5.comps.Group: a DNF object or None
         """
-        return self._base.comps.group_by_pattern(group_name)
+        for group in self._groups:
```
Sure, changed according to your suggestion for both `_get_group` and `_get_environment` methods.

M4rtinK reviewed (approved):
Looks good to me - left some comments for minor things. Thanks a lot! :)

pkratoch commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -147,15 +191,18 @@ def _create_base(cls):
+        # after configuration is updated, but before repositories
+        # are loaded or any query created.
+        # FIXME: Should we do that here?
+        #base.setup()
```
You probably had a pending review for some time, but submitted it only now? It's not there anymore.

pkratoch commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -483,10 +544,14 @@ def get_download_size(self):
-        self._base.reset(sack=True, repos=True, goal=True)
+
+        # FIXME: Reset sacks. Should we just drop the base?
+        # self._base.reset(sack=True, repos=True, goal=True)
```
Not there anymore.

pkratoch commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -495,25 +560,31 @@ def is_package_available(self, package_spec):
-            return False
+        #if not self._base.sack:
+        #    log.warning("There is no metadata about packages!")
+        #    return False
```
Not there anymore.

pkratoch commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -620,8 +663,9 @@ def set_download_location(self, path):
-            repo.pkgdir = path
+        # FIXME: Reimplement the assignment.
+        # for repo in self._base.repos.iter_enabled():
+        #    repo.pkgdir = path
```
Not there anymore.

pkratoch commented on tests/unit_tests/pyanaconda_tests/modules/payloads/payload/test_module_payload_dnf_manager.py:
```diff
@@ -21,17 +21,19 @@
+#from dnf.callback import PKG_SCRIPTLET, STATUS_FAILED, STATUS_OK
+#from dnf.comps import Comps, Environment, Group
+#from dnf.exceptions import DepsolveError, MarkingErrors, RepoError
+#from dnf.package import Package
```
Not there anymore.

pkratoch commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -676,27 +746,30 @@ def install_packages(self, callback, timeout=20):
             log.debug("The transaction has ended.")
-            base.close()  # Always close this base.
-            display.quit("DNF quit")
+            # base.close()  # Always close this base.
```
Yes, removed. Thanks.

pkratoch commented on pyanaconda/modules/payloads/payload/dnf/dnf_manager.py:
```diff
@@ -149,6 +149,7 @@ def _create_base(cls):
         config.cachedir = DNF_CACHE_DIR
         config.pluginconfpath = DNF_PLUGINCONF_DIR
         config.logdir = '/tmp/'
+        base.get_logger().add_logger(libdnf5.logger.create_file_logger(base, "dnf.log"))
```
Should be possible by inheriting from the libdnf Logger class and defining callbacks that use the Anaconda loggers. I think this could be resolved in a follow-up issue.

M4rtinK commented:
/kickstart-test --testtype smoke

M4rtinK commented:
/kickstart-test --testtype smoke

M4rtinK commented:
Bumped the DNF 5 version to 5.2.15 as agreed on with @pkratoch. This version is already in Rawhide & has the fix for the Anaconda GUI freezes. I will run the smoke tests & packaging tests next as final verification. :)

M4rtinK commented:
All green, time to merge. [Lets light this candle!](https://airandspace.si.edu/stories/editorial/alan-shepard-freedom-7) :D
